### PR TITLE
[Lang] MatrixType refactor part 2: add more ops

### PR DESCRIPTION
**Pull request by AD1024**
Issue: #5478 #5819

### CODE REVIEW

File: python/taichi/lang/matrix_ops_utils.py [hunk 1]
```diff
@@ -50,9 +76,64 @@ def assert_tensor(m, msg='not tensor type: {}'):
...
+
+def check_matmul(x, y):
+    assert_tensor(x, f'left hand side is not a matrix: {type(x)}')
+    assert_tensor(y, f'right hand side is not a matrix: {type(y)}')
+    x_shape = x.get_shape()
+    y_shape = y.get_shape()
+    if len(x_shape) == 1:
+        if len(y_shape) == 1:
+            return True, None
+        if x_shape[0] != y_shape[1]:
```
Comment (strongoier) [inline python/taichi/lang/matrix_ops_utils.py:134]:
```suggestion
        if x_shape[0] != y_shape[0]:
```
Let's pass CI tests - `y_shape[0]` is the number of rows of `y`.

File: python/taichi/lang/matrix_ops_utils.py [hunk 2]
```diff
@@ -50,9 +76,64 @@ def assert_tensor(m, msg='not tensor type: {}'):
...
+            return False, f'dimension mismatch between {x_shape} and {y_shape} for left multiplication'
+    else:
+        if x_shape[0] != y_shape[0]:
```
Comment (strongoier) [inline python/taichi/lang/matrix_ops_utils.py:137]:
```suggestion
        if x_shape[1] != y_shape[0]:
```
`x_shape[1]` is the number of columns of `x`.

File: python/taichi/lang/matrix_ops.py [hunk 3]
```diff
@@ -1,12 +1,305 @@
...
+                 (E(x, i + 1, 2, 4) * E(x, i + 3, 3, 4) -
+                  E(x, i + 3, 2, 4) * E(x, i + 1, 3, 4)) + E(x, i + 3, 1, 4) *
+                 (E(x, i + 1, 2, 4) * E(x, i + 2, 3, 4) -
+                  E(x, i + 2, 2, 4) * E(x, i + 1, 3, 4))))
+        return det
+    # unreachable
+    return None
+
+
+def _vector_transpose(v):
```
Comment (strongoier) [inline python/taichi/lang/matrix_ops.py:149]: As we've already handled matrix multiplication with vectors correctly, is it possible to make vector transpose always return a vector?

File: python/taichi/lang/matrix_ops_utils.py [hunk 4]
```diff
@@ -36,7 +36,33 @@ def check(*args, **kwargs):
...
+    return check
+
+
+def Or(f, g, msg=None):
+    def check(*args, **kwargs):
+        try:
+            do_check([f], *args, **kwargs)
+            return True, None
+        except TaichiCompilationError:
+            try:
```
Comment (jim19930609) [inline python/taichi/lang/matrix_ops_utils.py:60]: The other thing is `try-catch` has a maximum depth, and it's better only use `try-catch` in the outter-most function (`precondition` in your case). The inner functions should directly `raise` an Exception in some manner like: 
```
if xxxx:
    raise ...
```

**Review by jim19930609 — COMMENTED** [review]
I do think the decorators need another refactor. However, this shouldn't be blocking and I can help with that after all the attributes are finalized.

File: python/taichi/lang/matrix_ops.py (hunk 3 above)
Comment (AD1024) [reply python/taichi/lang/matrix_ops.py:149]: This issue is related to our change here: #6067 
Shall we un-support this operation (we definitely can make it a `v1.outer_product(v2)` though)

File: python/taichi/lang/matrix_ops_utils.py [hunk 5]
```diff
@@ -36,7 +36,33 @@ def check(*args, **kwargs):
...
+def foreach(*fns):
+    def check(args):
+        for x in args:
+            do_check(fns, x)
+        return True, None
+
+    return check
+
+
+def Or(f, g, msg=None):
```
[the remark this answers is not in the document]
Comment (AD1024) [reply python/taichi/lang/matrix_ops_utils.py:54]: This sounds good! Thanks!

**Review by jim19930609 — COMMENTED** [review]
Other parts LGTM.

**Review by strongoier — APPROVED** [review]
LGTM!

**Pull request merged**